### PR TITLE
[action] enable dynamic fee tx

### PR DESCRIPTION
**Pull request by dustinxie**
# Description
Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/rlp_tx.go [hunk 1]
```diff
@@ -125,30 +140,3 @@ func ExtractTypeSigPubkey(tx *types.Transaction) (iotextypes.Encoding, []byte, c
 	pubkey, err = crypto.RecoverPubkey(rawHash[:], sig)
 	return encoding, sig, pubkey, err
 }
-
-// ======================================
-// utility funcs to convert native action to eth tx
-// ======================================
-func toLegacyEthTx(ab TxCommon, act Action) (*types.Transaction, error) {
```
Comment (dustinxie) [inline action/rlp_tx.go:132]: moved to `envelope.ToEthTx()`

File: action/rlp_tx_test.go [hunk 2]
```diff
@@ -288,27 +288,25 @@ var (
...
+			"b30463d727248560e17d82764604fa64b2161f3bdb53434a6afd15fc64a42918",
+		},
+		{
+			"dynamicfee", 120, 21000, "1000000000", "0",
+			"0x3141df3f2e4415533bb6d6be2A351B2db9ee84EF",
+			_evmNetworkID,
+			iotextypes.Encoding_ETHEREUM_DYNAMICFEE,
+			nil,
+			"0dd507264fcc092c6430c248e48ac525e334e3699852e44451a376c4744e09af",
+		},
```
Comment (dustinxie) [inline action/rlp_tx_test.go:306]: test case for access list and dynamic fee tx

File: action/rlp_tx_test.go [hunk 3]
```diff
@@ -288,27 +288,25 @@ var (
...
-	tx := types.NewTx(&types.DynamicFeeTx{
-		To:        nil,
-		Nonce:     4,
-		Value:     big.NewInt(4),
-		Gas:       4,
-		GasTipCap: big.NewInt(44),
-		GasFeeCap: big.NewInt(1045),
-	})
-	_, _, _, err = ExtractTypeSigPubkey(tx)
-	require.ErrorIs(err, ErrNotSupported)
```
Comment (dustinxie) [inline action/rlp_tx_test.go:309]: these tx types are supported now, so the test become obsolete

File: action/sealedenvelope.go [hunk 4]
```diff
@@ -83,7 +84,8 @@ func (sealed *SealedEnvelope) calcHash() (hash.Hash256, error) {
 			return hash.ZeroHash256, ErrInvalidAct
 		}
 		return act.hash(), nil
-	case iotextypes.Encoding_ETHEREUM_EIP155, iotextypes.Encoding_ETHEREUM_UNPROTECTED:
+	case iotextypes.Encoding_ETHEREUM_EIP155, iotextypes.Encoding_ETHEREUM_UNPROTECTED, iotextypes.Encoding_ETHEREUM_ACCESSLIST,
```
Comment (envestcc) [inline action/sealedenvelope.go:87]: may use `Encoding_ETHEREUM_EIP155` to representing tx with eth-format

File: action/envelope.go [hunk 5]
```diff
@@ -217,16 +217,58 @@ func (elp *envelope) Action() Action { return elp.payload }
...
 // ToEthTx converts to Ethereum tx
 func (elp *envelope) ToEthTx(evmNetworkID uint32, encoding iotextypes.Encoding) (*types.Transaction, error) {
-	switch {
-	// TODO: handle dynamic fee tx and blob tx
-	case encoding == iotextypes.Encoding_IOTEX_PROTOBUF:
-		// treat native tx as EVM LegacyTx
-		fallthrough
-	case encoding == iotextypes.Encoding_ETHEREUM_EIP155 || encoding == iotextypes.Encoding_ETHEREUM_UNPROTECTED:
-		return toLegacyEthTx(elp.common, elp.Action())
+	tx, ok := elp.Action().(EthCompatibleAction)
```
Comment (CoderZhi) [inline action/envelope.go:220]: Add `ToEthTx(to, data)` to `EthCompatibleAction`, and the corresponding types will handle their own conversion logics by calling `return tx.ToEthTx(to, data)`

Comment (dustinxie) [reply action/envelope.go:220]: corresponding types -- do you mean `elp.Action() = transfer, execution, staking`? or the 4 types of actions `elp.common = legacy, accesslist, dynamicfee, blob`?
now the data are split into 2 parts:
1. `elp.Action()` provides 3 fields`Value, To, Data`
2. `elp.common` provides other fields like `Nonce, Gas, AccessList`, etc

so here the code is getting 3 fields from `elp.Action()`, and add to `elp.common` to construct the return tx

File: action/envelope.go [hunk 6]
```diff
@@ -217,58 +218,12 @@ func (elp *envelope) Action() Action { return elp.payload }
 
 // ToEthTx converts to Ethereum tx
 func (elp *envelope) ToEthTx(evmNetworkID uint32, encoding iotextypes.Encoding) (*types.Transaction, error) {
-	tx, ok := elp.Action().(EthCompatibleAction)
+	act, ok := elp.Action().(EthCompatibleAction)
```
Comment (dustinxie) [inline action/envelope.go:221]: still have `putpullresult` does not implement this, so for now this check is still needed
can remove this in another PR if we decide to implement `EthCompatibleAction` for `putpullresult`

**Review by envestcc — APPROVED** [review]

File: action/envelope.go [hunk 7]
```diff
@@ -221,58 +222,12 @@ func (elp *envelope) Action() Action { return elp.payload }
...
-			To:         to,
-			Value:      tx.Value(),
-			Data:       data,
-			AccessList: common.AccessList(),
-		}
-	default:
-		return nil, errors.Wrapf(ErrInvalidAct, "unsupported type = %v", common.Version())
-	}
-	return types.NewTx(txdata), nil
+	return elp.common.toEthTx(act)
```
Comment (CoderZhi) [inline action/envelope.go:230]:
```
        to, err := act.EthTo()
	if err != nil {
		return nil, err
	}
	data, err := act.EthData()
	if err != nil {
		return nil, err
	}
        return elp.common.toEthTx(to, data, act.Value())
```

**Review by CoderZhi — APPROVED** [review]

File: action/envelope.go (hunk 7 above)
Comment (dustinxie) [reply action/envelope.go]: changed to this

**Pull request merged**